### PR TITLE
Add runtime error in eval.py if yaml config is improperly formatted with extraneous or missing values

### PR DESCRIPTION
**Pull request by j316chuck**
## Description
This PR enables us to sanity check our eval yaml configuration files before we run the full eval pipeline. This enables us to catch errors in the YAML config prior to an eval run starting. If a yaml config is improperly formatted with extraneous or missing values, a runtime error will be thrown. 

## Unit Test:
Added unit tests to make sure we raise `omegaconf.errors` and/or warn users if the yaml is incorrectly formatted. Warnings are used if the parameter has an optional default value. 

## Integration Test:
**Test 1**: Model reaches 0.56 COPA accuracy before and after refactor (1 GPU) 
``` python 
  composer eval/eval.py \
  eval/yamls/hf_eval.yaml \
  icl_tasks=eval/yamls/copa.yaml \
  model_name_or_path=mpt-125m-hf
```

**Test 2**: Model gauntlet runs: `mcli run -f mcli/mcli-hf-eval.yaml --follow`

Main branch: `mcli logs -f all-eval-main-nphii5`
```
Ran mosaicml/mpt-7b-instruct eval in: 5852.91842341423 seconds
Printing gauntlet results for all models
| model_name               |   average |   world_knowledge |   commonsense_reasoning |   language_understanding |   symbolic_problem_solving |   reading_comprehension |
|:-------------------------|----------:|------------------:|------------------------:|-------------------------:|---------------------------:|------------------------:|
| mosaicml/mpt-7b-instruct |  0.354255 |          0.398764 |                0.415097 |                 0.371509 |                   0.171216 |                0.414691 |
Printing complete results for all models
| Category                 | Benchmark
```

chuck/add_yaml_check_eval branch: `mcli logs -f all-eval-chuck-branch-UqSCPR`
```
Ran mosaicml/mpt-7b-instruct eval in: 5858.392446279526 seconds
Printing gauntlet results for all models
| model_name               |   average |   world_knowledge |   commonsense_reasoning |   language_understanding |   symbolic_problem_solving |   reading_comprehension |
|:-------------------------|----------:|------------------:|------------------------:|-------------------------:|---------------------------:|------------------------:|
| mosaicml/mpt-7b-instruct |  0.354255 |          0.398764 |                0.415097 |                 0.371509 |                   0.171216 |                0.414691 |
Printing complete results for all models
| Category                 | Benchmark
```

## Issues Addressed
This PR chips away at this tech debt [issue](https://mosaicml.atlassian.net/browse/RESEARCH-717).

### CODE REVIEW

File: tests/test_eval_inputs.py [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        ]
+        old_cfg = copy.deepcopy(cfg)
+        for param in optional_params:
+            orig_value = cfg.pop(param, None)
+            updated_param = param + '-mispelling'
+            cfg[updated_param] = orig_value
+            with warnings.catch_warnings(record=True) as warning_list:
+                try:
+                    main(cfg)
+                except:
```
Comment (dakinggg) [inline tests/test_eval_inputs.py:69]: why the except here?

File: scripts/eval/eval.py [hunk 2]
```diff
@@ -120,20 +120,58 @@ def evaluate_model(model_cfg: DictConfig, cfg: DictConfig, run_name: str,
...
+    device_eval_batch_size: int = cfg.pop('device_eval_batch_size')
+    precision: str = cfg.pop('precision')
+
+    # Optional Evaluation Parameters with default values
+    seed: int = cfg.pop('seed', 17)
+    dist_timeout: Union[float, int] = cfg.pop('dist_timeout', 600.0)
+    default_run_name: str = os.environ.get('RUN_NAME', 'llm')
+    run_name: str = cfg.pop('run_name', default_run_name)
+    num_retries: int = cfg.pop('num_retries', 3)
+    loggers_cfg: Dict[str, Any] = cfg.pop('loggers', {}) or {}
```
Comment (dakinggg) [inline scripts/eval/eval.py:145]:
```suggestion
    loggers_cfg: Dict[str, Any] = cfg.pop('loggers', {})
```

File: scripts/eval/eval.py [hunk 3]
```diff
@@ -120,20 +120,58 @@ def evaluate_model(model_cfg: DictConfig, cfg: DictConfig, run_name: str,
...
+    cfg.pop('model_name_or_path', None)
```
Comment (dakinggg) [inline scripts/eval/eval.py:146]: Is this param unused?

**Review by dakinggg — COMMENTED** [review]
Could you please run the gauntlet before and after this pr (on a small model) to manually test the changes?

File: tests/test_eval_inputs.py (hunk 1 above)
Comment (j316chuck) [reply tests/test_eval_inputs.py:69]: we don't want to run the full eval run which takes a lot of time so we pass in an incomplete yaml that errors out early but raises the right warnings.

File: scripts/eval/eval.py (hunk 3 above)
Comment (dakinggg) [reply scripts/eval/eval.py:146]: Maybe haven't pushed this change?

**Review by dakinggg — COMMENTED** [review]
Looking good! Will approve once eval results are confirmed

**Comment by j316chuck** [comment]
added model gauntlet eval results that show that this change is non-breaking as results are the same before and after this PR.

**Review by dakinggg — APPROVED** [review]

**Pull request merged**